DRAFT - Open Liberty 21.0.0.11-beta blog (#1765)
OpenLiberty/blogs: posts/2021-10-05-beta-release-post.adoc

blog_description: Open Liberty 21.0.0.11-beta provides new configurable response headers, offering more granular control over response headers, as well as important changes to the Jakarta EE convenience features, with the removal of Jakarta EE 9.0 features in the transition to 9.1.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Configurable Response Headers and more exciting new features in Open Liberty 21.0.0.11-beta
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

Open Liberty 21.0.0.11-beta provides new configurable response headers, offering more granular control over response headers, as well as important changes to the Jakarta EE convenience features, with the removal of Jakarta EE 9.0 features in the transition to 9.1.

We have two beta packages for link:{url-about}[Open Liberty]:

> * <<allbeta, All Beta Features>>: a larger package that contains all Open Liberty beta features (including Jakarta EE 9 beta features) and GA features and functions.

Reviewer: ```suggestion
* <<allbeta, All Beta Features>>: a larger package that contains all Open Liberty beta features (including Jakarta EE 9.1 beta features) and GA features and functions.
```

> * <<jakarta, Jakarta EE 9 Beta Features>>: a lightweight package that contains only the Jakarta EE 9 features.

Reviewer: ```suggestion
* <<jakarta, Jakarta EE 9.1 Beta Features>>: a lightweight package that contains only the Jakarta EE 9.1 features.
```

                         httpsPort="9443"
                         headersRef="myHeadersID">
     </httpEndpoint>
 
     <httpEndpoint id="otherHttpEndpoint"
                         httpPort="9081"
                         httpsPort="9444"
                         headersRef="myHeadersID">
    </httpEndpoint> 
     
     <headers id="myHeadersID">
            <add>foo:bar</add>
            <add>foo:bar2</add>
            <set>customHeader:customValue</set>
            <setIfMissing>X-Forwarded-Proto:https</setIfMissing>
            <remove>Via</remove>
     </headers>  
----

 The `add` attribute allows multiple headers with the same name to be added to a response, similar to the HttpServletResponse's `addHeader` API. Similarly, the `set` attribute is analogous to the `setHeader` API, which sets a response header to the given name and value. This overwrites existing headers that share the same name. The `setIfMissing` attribute will only set the configured headers if they are not already present on the response. Lastly, the remove option will remove any response headers whose name matches a name defined by the configuration.

=== Try it now 

> To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 15, Java SE 11, or Java SE 8.

Reviewer: ```suggestion
To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 17, Java SE 11, or Java SE 8.
```

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
  <groupId>io.openliberty.beta</groupId>
  <artifactId>openliberty-runtime</artifactId>
  <version>20.0.0.11-beta</version>
  <type>pom</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty.beta', name: 'openliberty-runtime', version: '[20.0.0.11-beta,)'
}
----

Or take a look at our link:{url-prefix}/downloads/#runtime_betas[Downloads page].

[#jakarta]

> == Jakarta EE 9 Beta Features package

Reviewer: ```suggestion
== Jakarta EE 9.1 Beta Features package
```

As documented in the release of link:{url-prefix}/blog/2021/07/12/logging-format-21008-beta.html#JEE9_1[Open Liberty 21.0.0.8-beta], the convenience features `jakartaee-9.0`, `jakarteeClient-9.0` and `webProfile-9.0` have now been deleted in Open Liberty 21.0.0.11-beta, and completely replaced by `jakartaee-9.1`, `jakarteeClient-9.1` and `webProfile-9.1`.

> Enable the Jakarta EE 9 beta features in your app's `server.xml`. You can enable the individual features you want or you can just add the Jakarta EE 9 convenience feature to enable all of the Jakarta EE 9 beta features at once:

Reviewer: ```suggestion
Enable the Jakarta EE 9.1 beta features in your app's `server.xml`. You can enable the individual features you want or you can just add the Jakarta EE 9.1 convenience feature to enable all of the Jakarta EE 9.1 beta features at once:
```

[source, xml]
----
  <featureManager>
    <feature>jakartaee-9.1</feature>
  </featureManager>
----

> Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9 Web Profile beta features at once:

Reviewer: ```suggestion
Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9.1 Web Profile beta features at once:
```

[source, xml]
----
  <featureManager>
    <feature>webProfile-9.1</feature>
  </featureManager>
----

=== Try it now

> To try out these Jakarta EE 9 features on Open Liberty in a lightweight package, just update your build tools to pull the Open Liberty Jakarta EE 9 Beta Features package instead of the main release. The beta works with Java SE 15, Java SE 11, or Java SE 8.

Reviewer: ```suggestion
To try out these Jakarta EE 9.1 features on Open Liberty in a lightweight package, just update your build tools to pull the Open Liberty Jakarta EE 9.1 Beta Features package instead of the main release. The beta works with Java SE 17, Java SE 11, or Java SE 8.
```
